cpp/src/arrow/ipc/reader.cc

```diff
@@ -963,6 +1003,12 @@ class RecordBatchFileReaderImpl : public RecordBatchFileReader {
     // Get the schema and record any observed dictionaries
     RETURN_NOT_OK(UnpackSchemaMessage(footer_->schema(), options, &dictionary_memo_,
                                       &schema_, &out_schema_, &field_inclusion_mask_));
+    swap_endian_ = options.ensure_native_endian && !out_schema_->IsNativeEndianness();
+    if (swap_endian_) {
+      // create a new schema with native endianness before swapping endian in ArrayData
+      schema_ = schema_->WithNativeEndianness();
+      out_schema_ = out_schema_->WithNativeEndianness();
+    }
```

Comment:
this is basically the same as above, this suggests that something needs to be refactored